src/Admin/AdminHelper.php

```diff
@@ -223,7 +228,20 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
      */
     public function addNewInstance($object, FieldDescriptionInterface $fieldDescription)
     {
+        @trigger_error(sprintf(
+            'Method %s() is deprecated since sonata-project/admin-bundle 3.x. It will be removed in version 4.0.'
+            .' Use %s::addInstance() instead.',
+            __METHOD__,
+            __CLASS__
+        ), E_USER_DEPRECATED);
+
         $instance = $fieldDescription->getAssociationAdmin()->getNewInstance();
+
+        return self::addInstance($object, $fieldDescription, $instance);
+    }
+
+    public static function addInstance(object $object, FieldDescriptionInterface $fieldDescription, object $instance): object
```

Comment:
I don't remember what's our policy vis-à-vis prefixes but we should probably do something like this here:

```suggestion
    /**
     * @template T of object
     * @param T $instance
     * @return T
     */
    public static function addInstance(object $object, FieldDescriptionInterface $fieldDescription, object $instance): object
```